src/swarmpal/toolboxes/secs/dsecs_algorithms.py

```diff
@@ -0,0 +1,1895 @@
+            self.latB, self.grid.secs1Ddf.lat, self.rB, self.grid.Ri, 500
+        )
+
+        N1d = len(self.grid.secs1Ddf.lat)
+
+        y = self.Bpara  # measurement, parallel magnetic field
+
+        self.matBpara1D = (self.matBr1D.T * self.uvR).T + (
+            self.matBt1D.T * self.uvT
+        ).T  #
+
+        regmat = self.grid.secs1Ddf.diff2  # regularization
+        x = auto.sub_inversion(
+            self.matBpara1D, regmat, self.epsSVD1Ddf, self.alpha1Ddf, y
+        )
+        self.df1D = x
+        return x, y, self.matBpara1D
+
+    def fit2D_df(self):
+        """2D divergence-free fit for data."""
+        # Calculate B-matrices and form the field-aligned matrix
+        thetaB = (90 - self.latB) / 180 * np.pi
+        phiB = self.lonB / 180 * np.pi
+        theta2D = (90 - self.grid.secs2Ddf.magLat) / 180 * np.pi
+        phi2D = self.grid.secs2Ddf.magLon / 180 * np.pi
+
+        matBr2D, matBt2D, self.matBp2Ddf = SECS_2D_DivFree_magnetic(
+            thetaB, phiB, theta2D, phi2D, self.rB, self.grid.Ri
+        )
+        # N2d = np.size(self.grid.secs2D.lat)
+
+        self.matBpara2D = (
+            (matBr2D.T * self.uvR).T
+            + (matBt2D.T * self.uvT).T
+            + (self.matBp2Ddf.T * self.uvP).T
+        )
+
+        # Remove field explained by the 1D DF SECS (must have been fitted earlier).
+        Bpara2D = self.Bpara - self.matBpara1D @ self.df1D
+
+        regmat = self.grid.secs2Ddf.diff2lat2D  # regularization
+        self.df2D = auto.sub_inversion(
+            self.matBpara2D, regmat, self.epsSVD2D, self.alpha2D, Bpara2D
+        )
+
+        # Calculate the current produced by the 2D DF SECS.
+        thetaJ = (90 - self.grid.out.magLat.ravel()) / 180 * np.pi
+        phiJ = self.grid.out.magLon.ravel() / 180 * np.pi
+        matJt, matJp = SECS_2D_DivFree_vector(
+            thetaJ, phiJ, theta2D, phi2D, self.grid.Ri, self.grid.secs2Ddf.angle2D
+        )
+        df2dMagJt = np.reshape(matJt @ self.df2D, self.grid.out.magLat.shape)
+        df2dMagJp = np.reshape(matJp @ self.df2D, self.grid.out.magLat.shape)
+
+        # Calculate the magnetic field produced by the 2D DF SECS
+        ### still split into SwA and SwC
+        df2dBr = matBr2D @ self.df2D
+        df2dMagBt = matBt2D @ self.df2D
+        df2dMagBp = self.matBp2Ddf @ self.df2D
+        df2dBpara = self.matBpara2D @ self.df2D
+
+        return self.df2D, Bpara2D, self.matBpara2D, df2dBr
+
+    def fit1d_cf(self):
+        """1D curl-free fit for data."""
+        #        ####split data into hemispheres
+        # ind = np.nonzero(.............)
+        indN = np.argwhere(self.apexlats > 0)
+        indRN = np.argwhere(self.grid.out.magLat > 0)
+        indS = np.argwhere(self.apexlats <= 0)
+        indRS = np.argwhere(self.grid.out.magLat <= 0)
+
+        for ind, rind, gridhem in zip(
+            [indN, indS],
+            [indRN, indRS],
+            [self.grid.secs1DcfNorth, self.grid.secs1DcfSouth],
```

Comment:
`B905 `zip()` without an explicit `strict=` parameter.`
Just ignore this one for now. I found it's a feature only introduced in Python 3.10, and I maybe shouldn't have been so keen to update the pre-commit versions entirely!